internal/sourcemap/source_mapper.go

```diff
@@ -0,0 +1,313 @@
+	})
+
+	return &DocumentPositionMapper{
+		useCaseSensitiveFileNames: useCaseSensitiveFileNames,
+		sourceFileAbsolutePaths:   sourceFileAbsolutePaths,
+		sourceToSourceIndexMap:    sourceToSourceIndexMap,
+		generatedAbsoluteFilePath: generatedAbsoluteFilePath,
+		generatedMappings:         generatedMappings,
+		sourceMappings:            sourceMappings,
+	}
+}
+
+type DocumentPosition struct {
+	FileName string
+	Pos      int
+}
+
+func (d *DocumentPositionMapper) GetSourcePosition(loc *DocumentPosition) *DocumentPosition {
+	if d == nil {
+		return nil
+	}
+	if len(d.generatedMappings) == 0 {
+		return nil
+	}
+
+	targetIndex, _ := slices.BinarySearchFunc(d.generatedMappings, loc.Pos, func(m *MappedPosition, pos int) int {
+		return m.generatedPosition - pos
+	})
+
+	if targetIndex < 0 || targetIndex >= len(d.generatedMappings) {
+		return nil
+	}
+
+	mapping := d.generatedMappings[targetIndex]
+	if !mapping.isSourceMappedPosition() {
+		return nil
+	}
+
+	// Closest position
+	return &DocumentPosition{
+		FileName: d.sourceFileAbsolutePaths[mapping.sourceIndex],
+		Pos:      mapping.sourcePosition,
+	}
+}
+
+func (d *DocumentPositionMapper) GetGeneratedPosition(loc *DocumentPosition) *DocumentPosition {
+	if d == nil {
+		return nil
+	}
+	sourceIndex, ok := d.sourceToSourceIndexMap[tspath.GetCanonicalFileName(loc.FileName, d.useCaseSensitiveFileNames)]
+	if !ok {
+		return nil
+	}
+	if sourceIndex < 0 || int(sourceIndex) >= len(d.sourceMappings) {
+		return nil
+	}
+	sourceMappings := d.sourceMappings[sourceIndex]
+	targetIndex, _ := slices.BinarySearchFunc(sourceMappings, loc.Pos, func(m *SourceMappedPosition, pos int) int {
+		return m.sourcePosition - pos
+	})
+
+	if targetIndex < 0 || targetIndex >= len(sourceMappings) {
+		return nil
+	}
+
+	mapping := sourceMappings[targetIndex]
+	if mapping.sourceIndex != sourceIndex {
+		return nil
+	}
+
+	// Closest position
+	return &DocumentPosition{
+		FileName: d.generatedAbsoluteFilePath,
+		Pos:      mapping.generatedPosition,
+	}
+}
+
+func GetDocumentPositionMapper(host Host, generatedFileName string) *DocumentPositionMapper {
+	mapFileName := tryGetSourceMappingURL(host, generatedFileName)
+	if mapFileName != "" {
+		if base64Object, matched := tryParseBase64Url(mapFileName); matched {
+			if base64Object != "" {
+				if decoded, err := base64.StdEncoding.DecodeString(base64Object); err == nil {
+					return convertDocumentToSourceMapper(host, string(decoded), generatedFileName)
+				}
+			}
+			// Not a data URL we can parse, skip it
+			mapFileName = ""
+		}
+	}
+
+	var possibleMapLocations []string
+	if mapFileName != "" {
+		possibleMapLocations = append(possibleMapLocations, mapFileName)
+	}
+	possibleMapLocations = append(possibleMapLocations, generatedFileName+".map")
+	for _, location := range possibleMapLocations {
+		mapFileName := tspath.GetNormalizedAbsolutePath(location, tspath.GetDirectoryPath(generatedFileName))
+		if mapFileContents, ok := host.ReadFile(mapFileName); ok {
```

Comment:
problem with having these in LS and not project layer and watching these files is that "navigation": leads to discarding these files and having to reparse source maps etc and use to be bottle neck early on when it lived in LS. having it life of scriptInfo its associated with or till things change, meant we could avoid having to parse things again and again and that really helped with perf

Reply:
How could this work with the new snapshot architecture? I couldn't figure out exactly where this should live in Corsa. It used to live in `ProjectService` in Strada, and we don't have that anymore.

Reply:
it would live in session as thats kind of projectService now

Reply:
I am happy with a simpler solution for now even if it's slower. So if this PR looks good, then that's good in my eyes